backend/src/apiserver/metadata/metadata_store.go

```diff
@@ -0,0 +1,127 @@
+package metadata
+
+import (
+	"encoding/json"
+	"ml_metadata/metadata_store/mlmetadata"
+	mlpb "ml_metadata/proto/metadata_store_go_proto"
+	"strings"
+
+	argoWorkflow "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
+	"github.com/kubeflow/pipelines/backend/src/common/util"
+)
+
+// Store encapsulates a ML Metadata store.
+type Store struct {
+	mlmdStore *mlmetadata.Store
+}
+
+// NewStore creates a new Store, using mlmdStore as the backing ML Metadata
+// store.
+func NewStore(mlmdStore *mlmetadata.Store) *Store {
+	return &Store{mlmdStore: mlmdStore}
+}
+
+// RecordOutputArtifacts records metadata on artifacts as parsed from the Argo
+// output parameters in currentManifest. storedManifest represents the currently
+// stored manifest for the run with id runID, and is used to ensure metadata is
+// recorded at most once per artifact.
+func (s *Store) RecordOutputArtifacts(runID, storedManifest, currentManifest string) error {
+	storedWorkflow := &argoWorkflow.Workflow{}
+	if err := json.Unmarshal([]byte(storedManifest), storedWorkflow); err != nil {
+		return util.NewInternalServerError(err, "unmarshaling workflow failed")
+	}
+
+	currentWorkflow := &argoWorkflow.Workflow{}
+	if err := json.Unmarshal([]byte(currentManifest), currentWorkflow); err != nil {
+		return util.NewInternalServerError(err, "unmarshaling workflow failed")
+	}
+
+	completed := make(map[string]bool)
+	for _, n := range storedWorkflow.Status.Nodes {
+		if n.Completed() {
+			completed[n.ID] = true
```

Comment:
Why would that matter? We are operating within a specified runID, which is already unique right?

Reply:
It matters because if a step is run multiple times as part of a workflow, we end up with collisions. For instance, if the template is used twice within the same workflow we end up with the data being stored at: 
- workflow_UUID/node_ID for the first instance. 
- workflow_UUID/node_ID for the second instance. 

Hence we get a collision.

Reply:
I see that makes sense. I verified that this is the actual node id, not the step name. So it's actually unique.

Reply:
Thank you!